### PR TITLE
Parser change suggestions

### PR DESCRIPTION
**Pull request by MartinGC94**
# PR Summary
Moved the duplicate and positional argument checks to SemanticChecks.
Justification: This follows the existing conventions and improves the error recovery in the parsing.

Added tracking for successfully parsed label/colon tokens so the Ast extent for an incomplete statement can be more accurate.
Justification: This will make it easier to write the tab completion code due to the way the ast is analyzed there.

Replaced the "MissingArgumentAfterLabel" error with "MissingExpressionAfterToken" when there's no valid argument after a colon.
Justification: This seems more consistent with the normal error handling in the parser but this is obviously very subjective so feel free to discard this and similar suggestions.

Adjusted the error position for duplicate labels so the error is under the label rather than the label + expression.
Justification: Seems more accurate to me.

Adjusted the error position when there's a missing expression after the colon so it's after the colon instead of on the colon.
Justification: This matches the error reporting in other scenarios, like when there's a trailing comma.

Made the duplicate label check case insensitive.
Justification: Since you made it case insensitive in the binder there's no point in making it sensitive in the parser.

Removed a "Jordan" test Tag from the tests.
Justification: I assume you forgot to remove it yourself.

## PR Checklist

- [ ] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [ ] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [ ] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

File: test/powershell/Language/Parser/Parser.Tests.ps1 [hunk 1]
```diff
@@ -1486,17 +1461,31 @@ $c.Method(
...
         }
 
+        It "Fails to parse method invocation with label differing in case" {
+            $err = { ExecuteCommand '$c.Method(Arg: $value, arg: $value)' } | Should -Throw -ErrorId ParseException -PassThru
+            $pe = $err.Exception.InnerException
+            $pe | Should -BeOfType ([System.Management.Automation.ParseException])
+
+            $pe.Errors[0].ErrorId | Should -Be DuplicateArgumentLabel
+            $pe.Errors[0].Message | Should -Be 'Found duplicate argument name label "arg"'
+            $pe.ErrorRecord.InvocationInfo.PositionMessage | Should -Be (@(
```
Comment (jborean93) [inline test/powershell/Language/Parser/Parser.Tests.ps1]: Should this be a `-BeExactly` to ensure the error message contained `Arg:` on the first entry?

File: src/System.Management.Automation/engine/parser/Parser.cs [hunk 2]
```diff
@@ -7930,23 +7933,10 @@ private List<ExpressionAst> InvokeParamParenListRule(Token lParen, out IScriptEx
...
-                            }
-
-                            string nextTokenText = comma.Text;
-                            if (nextTokenText.Length == 1 && char.IsControl(nextTokenText[0]))
-                            {
-                                nextTokenText = string.Format("\\u{0:x4}", (short)nextTokenText[0]);
-                            }
-
-                            ReportIncompleteInput(After(argumentName.Extent),
-                                nameof(ParserStrings.MissingArgumentAfterLabel),
```
Comment (jborean93) [inline src/System.Management.Automation/engine/parser/Parser.cs:7946]: Does this entry need to be removed in the resources? I don't think it was used anywhere else.

**Review by jborean93 — COMMENTED** [review]
Thank you very much for the changes, they make sense to me. I wasn't aware of `SemanticChecks.cs` and moving those checks there is definitely easier to understand and maintain.

File: src/System.Management.Automation/engine/parser/Parser.cs (hunk 2 above)
Comment (MartinGC94) [reply src/System.Management.Automation/engine/parser/Parser.cs:7946]: You are right, I forgot to look in the resources. I've removed it now.

File: test/powershell/Language/Parser/Parser.Tests.ps1 (hunk 1 above)
Comment (MartinGC94) [reply test/powershell/Language/Parser/Parser.Tests.ps1]: Yes, tbh I barely looked at the test as I just copied the one you had made for the duplicate check and changed the casing of the input.

Comment (jborean93) [reply test/powershell/Language/Parser/Parser.Tests.ps1]: All good, I appreciate you fixing it up anyway!

**Comment by jborean93** [comment]
Thank you very much for your suggestions.

**Pull request merged**